### PR TITLE
Experiment "Print a page table" of lab page tables

### PR DESCRIPTION
**Pull request by LittleYe233**
Formats some original source files via clang-format.

Completes all related source code and answers additional questions.

Finishes the whole experiment.

This PR is based on #10 which is awaiting reviews.

### CODE REVIEW

**Review by inclyc — COMMENTED** [review]
> Finish the experiment "Print a page table" of lab page tables

Re-title this.

> [feat(pgtbl/speedup): complete code](https://github.com/inclyc/Sophia-xv6-labs-2021/pull/11/commits/348a7ebdcdc37bcc3ab03c7c4ede32ddd9c9bff4)

Be more specific in writing commit messages. For example, say that you are addressing "vmprint" lab.

> [feat(pgtbl/vmprint): complete additional questions](https://github.com/inclyc/Sophia-xv6-labs-2021/pull/11/commits/e19e2f45b2316e95386b4d803968c145f8584daa)

Are you completing everything? Just say answer the questions.

File: answers-pgtbl.txt [hunk 1]
```diff
@@ -8,3 +8,27 @@ By storing the pointer to the parent process in `usyscall` struct, the child pro
...
+
+Data and text of a process.
+
+* What is in page 2?
+
+A stack containing the information of command-line arguments.
+
+* Page 1 in user mode
+
+No, because this page does not have the `PTE_U` flag.
```
Comment (inclyc) [inline answers-pgtbl.txt:30]: is not attributed.

File: answers-pgtbl.txt [hunk 2]
```diff
@@ -8,3 +8,27 @@ By storing the pointer to the parent process in `usyscall` struct, the child pro
...
+
+* What does the third to last page contain?
+
+1) Heap, 2) a shared page related to the current process (`struct usyscall`) for system call optimization, 3) trapframe, 4) trampoline.
```
Comment (inclyc) [inline answers-pgtbl.txt:34]: We do not say "Heap" in kernel development.

File: kernel/vm.c [hunk 3]
```diff
@@ -395,3 +395,28 @@ int copyinstr(pagetable_t pagetable, char *dst, uint64 srcva, uint64 max) {
...
+      if ((pte & (PTE_R | PTE_W | PTE_X)) == 0) {
+        // This PTE points to a lower-level page table.
+        vmprint_with_depth((pagetable_t)pa, depths + 1);
+      }
+    }
+  }
+}
+
+// Internal depth indicators
+const char *VMPRINT_DEPTHS[] = {" ..", " .. ..", " .. .. .."};
```
Comment (inclyc) [inline kernel/vm.c:416]: What will happen if RISC-V specifies that there is a deeper page table? Or, someday you want to migrate your project in more architectures?

File: kernel/vm.c [hunk 4]
```diff
@@ -395,3 +395,28 @@ int copyinstr(pagetable_t pagetable, char *dst, uint64 srcva, uint64 max) {
     return -1;
   }
 }
+
+// Recursively print the page table in a tree view with depth indicators.
+void vmprint_with_depth(pagetable_t pagetable, const char *depths[]) {
+  for (int i = 0; i < 512; ++i) {
+    pte_t pte = pagetable[i];
+    if (pte & PTE_V) {
+      // This PTE is valid.
```
Comment (inclyc) [inline kernel/vm.c:404]: Drop this comment, `pte & PTE_V` is self-explained.

File: kernel/defs.h [hunk 5]
```diff
@@ -10,177 +10,177 @@ struct stat;
 struct superblock;
 
 // bio.c
-void            binit(void);
-struct buf*     bread(uint, uint);
-void            brelse(struct buf*);
-void            bwrite(struct buf*);
-void            bpin(struct buf*);
-void            bunpin(struct buf*);
+void binit(void);
```
Comment (inclyc) [inline kernel/defs.h:13]: mark this commit NFC.

File: answers-pgtbl.txt (hunk 2 above)
Comment (LittleYe233) [reply answers-pgtbl.txt:34]: I wonder why not. I used this word from the figure of a process's user address space in the xv6 textbook, and I am not sure which word is better here.

Comment (inclyc) [reply answers-pgtbl.txt:34]: Fine, let's keep this word.

**Review by inclyc — APPROVED** [review]

**Pull request merged**